certbot/certbot/compat/misc.py

```diff
@@ -116,23 +116,25 @@ def underscores_for_unsupported_characters_in_path(path):
     return drive + tail.replace(':', '_')
 
 
-def execute_command(cmd_name, shell_cmd):
-    # type: (str, str) -> Tuple[str, str]
+def execute_command(cmd_name, shell_cmd, env=None):
+    # type: (str, str, dict) -> Tuple[str, str]
     """
     Run a command:
         - on Linux command will be run by the standard shell selected with Popen(shell=True)
         - on Windows command will be run in a Powershell shell
 
     :param str cmd_name: the user facing name of the hook being run
     :param str shell_cmd: shell command to execute
+    :param dict env: environ to pass into Popen on Linux
```

Comment:
Quietly ignoring this parameter on Windows is a little awkward to me and since this is part of the public API, if we have a need to set the environment while running a command on Windows in the future, we'd either have to make a new function/parameter or technically make an API change.

To avoid this especially since `env` on Windows should be unmodified by stripping out snap paths, I think we should just pass in `env` to subprocess on Windows here too.

Reply:
I had that initially, but then I was thinking, what if somehow we accidentally get past the `if`? If we split on `:` on Windows, I could see things going terribly wrong. That's a good point about the API change though.

Reply:
That's a good point too. I think it's unlikely, but I could imagine someone using the environment variable `SNAP` for something else and things going horribly wrong, especially if they were a Windows user.

If we wanted to make the decision of whether or not to modify paths more robust, probably my preferred approach would be to set an environment variable like `CERTBOT_SNAPPED=1` in `certbot.wrapper` and check for both of these variables before modifying the environment. My thinking is `CERTBOT_SNAPPED` is extremely specific so there are unlikely to be conflicts. (If someone does show up complaining about conflicts, we can complain to them about improper use of our trademark :stuck_out_tongue_closed_eyes: )

Another idea would be to check that we're not on Windows using `os.name` or `sys.platform`.

Reply:
Wow, that is uh certainly a novel approach to trademark. I'll look at `CERTBOT_SNAPPED` and `os.name` or `sys.platform` and see how I feel about that.

Reply:
(I should probably add that I am not a lawyer :smile: )